portfolio/src/main/webapp/script.js

```diff
@@ -16,45 +16,81 @@
+/** @type {{imgSrc: string, description: string, altText: string }}  */
+const GALLERY_ITEMS = [
+  {
+    imgSrc: "food_picture.jpg",
+    description: "My first time eating at Toshi Tiger near the university. The flavor changed my life forever.",
+    altText: "A picture of a table with a plate of noodles with meat and vegetables on its top with a plate of bread with shrimp inside it."
+  },
+  {
+    imgSrc: "canada_picture.jpg",
+    description: "A family vacation in Canada. Who would have thought that they have amazing views.",
+    altText: "An amazing view of the Parlament of Canada and Ottawa River taken from Major's Hill Park."
+  },
+  {
+    imgSrc: "monterrey_picture.JPG",
+    description: "An amazing sunset over my city Monterrey!",
+    altText: "A beautiful sunset with a deep blue sky with a mix of orange flaming rays of light over Monterrey, Mexico"
+  },
+  {
+    imgSrc: "building_picture.JPG",
+    description: "Really appreciate good architecture.",
+    altText: "A gothic styled building painted of orange with green roofs on Quebec City"
+  },
+  {
+    imgSrc: "firstday_picture.jpg",
+    description: "My first day as a Google STEP intern!",
+    altText: "Me smiling with a top knot, wearing a purple shirt with the label 'Hacker', with my Google's Chromebook in front of me"
+  },
+  {
+    imgSrc: "highview_picture.jpg",
+    description: "I get really excited when I have views like this.",
+    altText: "A view of Montreal from the top of the Olympic Stadium"
+  }
+];
+
+const TIME_INTERVAL = 10000; // Constant time of 10 seconds
+let galleryItemsStack = []; // Stack to keep track of displayed images
+let randomImageEvent = false; // Bool to track onClick events
+let intervalState = null; // Variable to store the setInterval() status
 
 function getRandomImage() {
 
+  // Change statuks of the event on every click
+  randomImageEvent = !randomImageEvent
+  const randomButtonContainer = document.querySelector('.random_picture_button');
+
+  if(randomImageEvent) {
+
+    //Start the random image cycler
+    intervalState= setInterval(setRandomImage,TIME_INTERVAL);
+    randomButtonContainer.textContent = "Click to stop the random image cycler";
+  } else {
+
+    //Stop the random image cycler
+    randomButtonContainer.textContent = "Click to start the random image cycler";
+    galleryItemsStack = [];
+    clearInterval(intervalState);
+    intervalState = null;
+  }
+}
+
+function setRandomImage() {
+
+  // Restart the stack if all images have been displayed in a cycle
+  if(galleryItemsStack.length === GALLERY_ITEMS.length) galleryItemsStack = [];
+
   // Pick a random gallery item
-  const singleGalleryItem = GALLERY_ITEMS[Math.floor(Math.random() * GALLERY_ITEMS.length)];
+  let singleGalleryItem = GALLERY_ITEMS[Math.floor(Math.random() * GALLERY_ITEMS.length)];
+
+  // Look for another image if the one has already been displayed in the cycle
+  while(galleryItemsStack.indexOf(singleGalleryItem) > -1) {
```

Comment:
Correct me if I'm wrong, but this is technically unbounded right? I mean it would never happen, but you could possibly get stuck in this loop forever?

See if you can play around with other data structures, or perhaps pre-generating a scrambled array each time you need to "reset", to find a solution that will allow you to say for certain that it will only take x operations to find the next image.

Reply:
Actually I don' know if it can get stuck forever, but it the number of operations between images can be totally different in every run so I do get the point. Are you suggesting to shuffle the GALLERY_ITEMS array every cycle and iterate over it instead of randomly looking for available indexes?

Reply:
Yes, something to that effect.

Ideally you do not want to be picking a random variable, and then checking that that random variable meets some condition (in this case that the random variable is not an index of something already added to the stack). This is because you have no guarantee that a random variable will ever meet your condition, and you have no control over the randomness.

Instead I would suggest something closer to "shuffling" the array. You could select two indices at random, and then flip the elements at those indices in the array. You could do that X times (where X is some number that ensures your array is sufficiently shuffled), and sure occasionally you'd be swapping an element with itself when your random numbers collide, but with a sufficiently high X it wouldn't matter, and you would be able to say for certain how many operations this will take.

Apparently there is an algorithm for this, although I have not used it by name: https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle

No need to implement it exactly, but you could use it for inspiration.

Reply:
Looks good Mark, thank you for the suggestion and the article! I looked at the algorithm and it doesn't seem much of a complication, so I'll try it to see how it works